### DISCUSSION TITLE
关于find_package返回值

### DISCUSSION THREAD

**Discussion by x1244** [xmake-io/xmake#2841] [Q&A]
### Xmake 版本

2.7.1

### 操作系统版本和架构

Windows10

### 描述问题

按手册说明，`find_package`如果能查找到包，则返回一个包含所有包信息的table，如果失败返回nil。

我这里的`find_package`在包存在的情况下，不会返回任何值。但是在包存在的情况下，`add_packages`可以将包正确添加并能够使用。

请问是不是`find_package`接口改变了，还是`add_packages`不再通过`find_package`来添加包了？

又看Issues说`find_package`是很老的接口了，那如何能在脚本域中获取包的信息呢？

### 期待的结果

有个私有包，只是想打印出来看一看是不是在正确的地方。

### 工程配置

add_requires("spdlog")
target("pkg")
    set_kind("binary")
    add_packages("spdlog")
    add_files("src/*.cpp")
    before_build(function (target)
        import("lib.detect.find_package")
        local pkg = find_package("spdlog")
        print(pkg or "nil")
    end)

### 附加信息和错误日志

PS tx> xmake build --rebuild
--> nil
[ 20%]: cache compiling.release src\backward.cpp
[ 20%]: cache compiling.release src\main.cpp
[ 60%]: linking.release pkg.exe
[100%]: build ok!

PS tx> xrepo info spdlog
The package info of project:
    require(spdlog):
      -> description: Fast C++ logging library.
      -> version: v1.10.0
      -> urls:
         -> https://github.com/xmake-mirror/build-artifacts/releases/download/spdlog-v1.10.0/windows-x64-vc142-6bf6da47960848619cd072b319304dad.7z
            -> c4f93b6c3803fe9e26c097469560e398188b0c025bcca59a687343133068115e
      -> repo: xmake-repo https://github.com/xmake-io/xmake-repo.git master
      -> cachedir: C:\Users\conlay\AppData\Roaming\.xmake\cache\packages\2209\s\spdlog\v1.10.0
      -> installdir: C:\Users\conlay\AppData\Roaming\.xmake\packages\s\spdlog\v1.10.0\6bf6da47960848619cd072b319304dad
      -> searchdirs:
      -> searchnames: spdlog-v1.10.0.7z, windows-x64-vc142-6bf6da47960848619cd072b319304dad.7z
      -> fetchinfo: v1.10.0
          -> version: v1.10.0
          -> sysincludedirs: C:\Users\conlay\AppData\Roaming\.xmake\packages\s\spdlog\v1.10.0\6bf6da47960848619cd072b319304dad\include
      -> platforms: all
      -> requires:
         -> plat: windows
         -> arch: x64
         -> configs:
            -> noexcept: false
            -> header_only: true
            -> pic: true
            -> fmt_external: false
            -> debug: false
            -> vs_runtime: MT
            -> shared: false
      -> configs:
         -> header_only: Use header only (default: true)
         -> fmt_external: Use external fmt library instead of bundled (default: false)
         -> noexcept: Compile with -fno-exceptions. Call abort() on any spdlog exceptions (default: false)
         -> shared: Build shared library. (default: false)
      -> configs (builtin):
         -> debug: Enable debug symbols. (default: false)
         -> pic: Enable the position independent code. (default: true)
         -> lto: Enable the link-time build optimization. (type: boolean)
         -> vs_runtime: Set vs compiler runtime.
            -> values: {"MT","MTd","MD","MDd"}
         -> toolchains: Set package toolchains only for cross-compilation.
         -> cflags: Set the C compiler flags.
         -> cxflags: Set the C/C++ compiler flags.
         -> cxxflags: Set the C++ compiler flags.
         -> asflags: Set the assembler flags.

**Comment by waruqi** [comment]
集成包 不需要 find_package， 这个是 deprecated 接口，目前仅供内部使用。文档后期关于这块的 我会去掉。。

请用 add_requires/add_packages 集成包

**Comment by x1244** [comment]
我也知道集成包不需要`find_package`，用add_requires/add_packages 可以完成任务。
我还想问的是有没有办法在脚本域获得一个包的相关信息？
就像原来`find_package`提供的功能一样。

**Reply by waruqi** [reply]
```lua
before_build(function (target)
    local zlib = target:pkg("zlib")
    if zlib then
          print(zlib:get("links"), zlib:get("includedirs"))
    end
end)
```

**Reply by waruqi** [reply]
> 有个私有包，只是想打印出来看一看是不是在正确的地方。

只是调试，直接 `xmake -v` 看 `-I/-L` 不就好了，或者看 `xmake require --info zlib` 看详情

**Comment by x1244** [comment]
明白了，十分感谢。
但是在你给的示例中有个错误，
print(zlib:get("links"), zlib:get("includedirs"))
通过打印和调试
`includedirs`应该是`sysincludedirs`
这样才能正确获得输出。

**Reply by waruqi** [reply]
我随手敲的，只是举例

**Comment by xixioba** [comment]
`add_requires`添加的包有些include不对，比如conda::pcl 1.13.0，需要按版本添加一段路径，得手动修一下，不知道还有没有更好的方法直接获取`add_requires`的table了或者可以修复conda::pcl的问题了🤔

**Reply by waruqi** [reply]
提 pr 过来改进 https://github.com/xmake-io/xmake/blob/master/xmake/modules/package/manager/conda/find_package.lua

或者改用 xmake-repo 仓库的包

**Reply by xixioba** [reply]
研究了一下，conda的find_package.lua没办法完美适配，有个更简单的想法，比如
```
add_requires("conda::pcl 1.13.0", {alias = "pcl"})
target("test_pcl")
    set_languages("c++17")
    add_files("test.cc")
    add_packages("pcl")
    before_build(function (target)
    local pkg = target:pkg("pcl")
    if pkg then
        -- print(pkg:get("links"), pkg:get("sysincludedirs"))
        pclpath = pkg:get("sysincludedirs").."/pcl-1.13"
        print("Add Conda PCL install Path: ", pclpath)
        add_includedirs("$(pclpath)")     <--- here
        end
    end)
```
有没有办法做类似的操作，或者可以修改`sysincludedirs`，那样就可以任意适配了

**Reply by xixioba** [reply]
🥲突然发现有[target:set](https://xmake.io/#/manual/target_instance?id=targetset)，对我而言可以满足需求
替换为即可:
`pkg:set("sysincludedirs",pclpath)`
